src/realm/util/encrypted_file_mapping.cpp

```diff
@@ -279,6 +258,42 @@ bool AESCryptor::check_hmac(const void* src, size_t len, const uint8_t* hmac) co
     return result == 0;
 }
 
+std::map<size_t, IVRefreshState> AESCryptor::refresh_ivs(FileDesc fd, off_t data_pos,
+                                                         size_t page_ndx_in_file_expected)
+{
+    std::map<size_t, IVRefreshState> page_states;
+    std::vector<iv_table> ivs_before;
+    ivs_before.reserve(blocks_per_metadata_block);
+
+    REALM_ASSERT(!int_cast_has_overflow<size_t>(data_pos));
+    size_t data_pos_casted = size_t(data_pos);
+    // the call to get_iv_table() below reads in all ivs in a chunk with size = blocks_per_metadata_block
+    // so we will know if any page in this chunk has changed
+    const size_t page_ndx_of_refresh_chunk_start =
+        ((data_pos_casted / page_size()) / blocks_per_metadata_block) * blocks_per_metadata_block;
+    REALM_ASSERT_EX(page_ndx_of_refresh_chunk_start + blocks_per_metadata_block <= m_iv_buffer.size(),
+                    page_ndx_of_refresh_chunk_start, blocks_per_metadata_block, m_iv_buffer.size());
+    REALM_ASSERT_EX(page_ndx_in_file_expected >= page_ndx_of_refresh_chunk_start &&
+                        page_ndx_in_file_expected < page_ndx_of_refresh_chunk_start + blocks_per_metadata_block,
+                    page_ndx_in_file_expected, page_ndx_of_refresh_chunk_start, blocks_per_metadata_block);
+    for (size_t i = page_ndx_of_refresh_chunk_start; i < page_ndx_of_refresh_chunk_start + blocks_per_metadata_block;
+         ++i) {
+        page_states[i] = IVRefreshState::UpToDate;
+        ivs_before.push_back(m_iv_buffer[i]);
+    }
+
+    get_iv_table(fd, data_pos, IVLookupMode::Refetch);
+
+    for (size_t i = page_ndx_of_refresh_chunk_start; i < page_ndx_of_refresh_chunk_start + blocks_per_metadata_block;
+         ++i) {
+        if (ivs_before[i - page_ndx_of_refresh_chunk_start] != m_iv_buffer[i]) {
+            REALM_ASSERT_EX(page_states.count(i) == 1, i, m_iv_buffer.size());
+            page_states[i] = IVRefreshState::RequiresRefresh;
+        }
+    }
```

Comment:
I think introducing a variable for the end page index rather than recalculating it each time can make this logic a bit clearer:

```suggestion
const size_t page_ndx_refresh_start =
        ((data_pos_casted / page_size()) / blocks_per_metadata_block) * blocks_per_metadata_block;
    const size_t page_ndx_refresh_end = page_ndx_refresh_start + blocks_per_metadata_block;
    REALM_ASSERT_EX(page_ndx_refresh_end <= m_iv_buffer.size(), page_ndx_refresh_start, blocks_per_metadata_block,
                    m_iv_buffer.size());
    REALM_ASSERT_EX(page_ndx_in_file_expected >= page_ndx_refresh_start &&
                        page_ndx_in_file_expected < page_ndx_refresh_end,
                    page_ndx_in_file_expected, page_ndx_refresh_start, page_ndx_refresh_end);
    for (size_t i = page_ndx_refresh_start; i < page_ndx_refresh_end; ++i) {
        page_states[i] = IVRefreshState::UpToDate;
        ivs_before.push_back(m_iv_buffer[i]);
    }

    get_iv_table(fd, data_pos, IVLookupMode::Refetch);

    for (size_t i = page_ndx_refresh_start; i < page_ndx_refresh_end; ++i) {
        if (ivs_before[i - page_ndx_refresh_start] != m_iv_buffer[i]) {
            REALM_ASSERT_EX(page_states.count(i) == 1, i, m_iv_buffer.size());
            page_states[i] = IVRefreshState::RequiresRefresh;
        }
    }
```